File: test/core/utils/shop_time_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:maki_mobile_pos/core/utils/shop_time.dart';
import 'package:maki_mobile_pos/core/utils/shop_timezones.dart';

void main() {
  const ph = 480; // Asia/Manila
  const est = -300; // UTC-5, a "device in the US" stand-in

  group('shopTimeOf', () {
    test('shifts an instant into shop wall time', () {
      final instant = DateTime.utc(2026, 8, 26, 15, 30);
      final wall = shopTimeOf(instant, ph);
      expect(wall.year, 2026);
      expect(wall.month, 8);
      expect(wall.day, 26);
      expect(wall.hour, 23);
      expect(wall.minute, 30);
    });

    test('is independent of the DateTime the caller hands in', () {
      // Same instant expressed two ways must produce the same wall time.
      final utc = DateTime.utc(2026, 8, 26, 15, 30);
      final shifted = utc.add(const Duration(hours: 3)).subtract(const Duration(hours: 3));
      expect(shopTimeOf(shifted, ph), shopTimeOf(utc, ph));
    });

    test('crosses the day boundary at shop midnight, not device midnight', () {
      // 16:00 UTC on the 25th is 00:00 of the 26th in PH.
      final instant = DateTime.utc(2026, 8, 25, 16, 0);
      expect(shopDayInt(instant, ph), 20260826);
      // The same instant is still the 25th for a UTC-5 device.
      expect(shopDayInt(instant, est), 20260825);
    });
  });

  group('instantOf', () {
    test('round-trips with shopTimeOf', () {
      final instant = DateTime.utc(2026, 8, 26, 15, 30, 45, 123);
      expect(instantOf(shopTimeOf(instant, ph), ph), instant);
      expect(instantOf(shopTimeOf(instant, est), est), instant);
    });

    test('returns a UTC instant for a shop wall time', () {
      final wall = shopWall(2026, 8, 26); // shop midnight
      final instant = instantOf(wall, ph);
      expect(instant.isUtc, isTrue);
      expect(instant, DateTime.utc(2026, 8, 25, 16, 0));
    });
  });

  group('shopDayInt', () {
    test('matches the rules yyyymmdd shape', () {
      expect(shopDayInt(DateTime.utc(2026, 1, 2, 4, 0), ph), 20260102);
    });

    test('handles year boundaries', () {
      // 16:00 UTC Dec 31 is 00:00 Jan 1 in PH.
      expect(shopDayInt(DateTime.utc(2026, 12, 31, 16, 0), ph), 20270101);
    });

    test('handles a negative offset', () {
      // 02:00 UTC Jan 1 is 21:00 Dec 31 at UTC-5.
      expect(shopDayInt(DateTime.utc(2027, 1, 1, 2, 0), est), 20261231);
    });
  });

  group('shopDateOf', () {
    test('truncates to shop midnight and stays a wall value', () {
      final d = shopDateOf(DateTime.utc(2026, 8, 26, 15, 30), ph);
      expect(d, shopWall(2026, 8, 26));
      expect(d.isUtc, isTrue);
      expect(d.hour, 0);
    });
  });

  group('ShopTimeConfig + extensions', () {
    setUp(() => ShopTimeConfig.offsetMinutes = kDefaultShopOffsetMinutes);

    test('defaults to Asia/Manila', () {
      expect(ShopTimeConfig.offsetMinutes, 480);
      expect(ShopTimeConfig.timezoneId, 'Asia/Manila');
    });

    test('inShopTime uses the ambient offset', () {
      ShopTimeConfig.offsetMinutes = est;
      expect(DateTime.utc(2026, 8, 26, 15, 30).inShopTime.hour, 10);
    });

    test('asShopInstant inverts inShopTime', () {
      final instant = DateTime.utc(2026, 8, 26, 15, 30);
      expect(instant.inShopTime.asShopInstant, instant);
    });
  });

  group('kShopTimezones', () {
    test('contains Asia/Manila at +480 as the default', () {
      final manila = shopTimezoneById(kDefaultShopTimezoneId);
      expect(manila, isNotNull);
      expect(manila!.offsetMinutes, kDefaultShopOffsetMinutes);
    });

    test('every offset is within the supported range', () {
      for (final tz in kShopTimezones) {
        expect(tz.offsetMinutes, inInclusiveRange(-720, 840));
      }
    });

    test('ids are unique', () {
      final ids = kShopTimezones.map((t) => t.id).toSet();
      expect(ids.length, kShopTimezones.length);
    });

    test('unknown id returns null', () {
      expect(shopTimezoneById('Mars/Olympus'), isNull);
    });
  });
}

File: web_admin/src/domain/time/shopTime.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import {
  DEFAULT_SHOP_OFFSET_MINUTES,
  formatShopDateTime,
  DEFAULT_SHOP_TIMEZONE_ID,
  getAmbientShopTimezone,
  instantOf,
  setAmbientShopTimezone,
  shopDateKey,
  shopDayInt,
  shopEndOfDay,
  shopIsoDate,
  shopOffsetMinutes,
  shopStartOfDay,
  shopTimeOf,
  shopWall,
} from './shopTime';
import { SHOP_TIMEZONES, shopTimezoneById } from './shopTimezones';

const PH = 480;
const EST = -300;

describe('shopTimeOf', () => {
  it('shifts an instant into shop wall time (read with getUTC*)', () => {
    const wall = shopTimeOf(new Date(Date.UTC(2026, 7, 26, 15, 30)), PH);
    expect(wall.getUTCFullYear()).toBe(2026);
    expect(wall.getUTCMonth()).toBe(7);
    expect(wall.getUTCDate()).toBe(26);
    expect(wall.getUTCHours()).toBe(23);
  });
});

describe('instantOf', () => {
  it('round-trips with shopTimeOf', () => {
    const instant = new Date(Date.UTC(2026, 7, 26, 15, 30, 45, 123));
    expect(instantOf(shopTimeOf(instant, PH), PH).getTime()).toBe(instant.getTime());
    expect(instantOf(shopTimeOf(instant, EST), EST).getTime()).toBe(instant.getTime());
  });

  it('maps shop midnight to the right instant', () => {
    expect(instantOf(shopWall(2026, 8, 26), PH).toISOString()).toBe('2026-08-25T16:00:00.000Z');
  });
});

describe('shopDayInt', () => {
  it('matches the rules yyyymmdd shape', () => {
    expect(shopDayInt(new Date(Date.UTC(2026, 7, 26, 5, 0)), PH)).toBe(20260826);
  });

  it('crosses the day boundary at shop midnight', () => {
    const instant = new Date(Date.UTC(2026, 7, 25, 16, 0));
    expect(shopDayInt(instant, PH)).toBe(20260826);
    expect(shopDayInt(instant, EST)).toBe(20260825);
  });

  it('handles a year boundary', () => {
    expect(shopDayInt(new Date(Date.UTC(2026, 11, 31, 16, 0)), PH)).toBe(20270101);
  });
});

describe('day bounds', () => {
  it('shopStartOfDay returns the instant of shop midnight', () => {
    const i = new Date(Date.UTC(2026, 7, 26, 5, 0));
    expect(shopStartOfDay(i, PH).toISOString()).toBe('2026-08-25T16:00:00.000Z');
  });

  it('shopEndOfDay returns the last millisecond of the shop day', () => {
    const i = new Date(Date.UTC(2026, 7, 26, 5, 0));
    expect(shopEndOfDay(i, PH).toISOString()).toBe('2026-08-26T15:59:59.999Z');
  });

  it('the bounds span exactly one day', () => {
    const i = new Date(Date.UTC(2026, 7, 26, 5, 0));
    expect(shopEndOfDay(i, PH).getTime() - shopStartOfDay(i, PH).getTime()).toBe(86_400_000 - 1);
  });
});

describe('key formatting', () => {
  it('shopDateKey is zero-padded YYYYMMDD', () => {
    expect(shopDateKey(new Date(Date.UTC(2026, 0, 2, 5, 0)), PH)).toBe('20260102');
  });

  it('shopIsoDate is zero-padded yyyy-MM-dd', () => {
    expect(shopIsoDate(new Date(Date.UTC(2026, 0, 2, 5, 0)), PH)).toBe('2026-01-02');
  });
});

describe('ambient timezone', () => {
  beforeEach(() =>
    setAmbientShopTimezone({
      timezoneId: DEFAULT_SHOP_TIMEZONE_ID,
      offsetMinutes: DEFAULT_SHOP_OFFSET_MINUTES,
    }),
  );

  it('defaults to Asia/Manila', () => {
    expect(shopOffsetMinutes()).toBe(480);
    expect(getAmbientShopTimezone().timezoneId).toBe('Asia/Manila');
  });

  it('is used when no offset is passed', () => {
    setAmbientShopTimezone({ timezoneId: 'Asia/Tokyo', offsetMinutes: 540 });
    expect(shopDayInt(new Date(Date.UTC(2026, 7, 25, 15, 30)))).toBe(20260826);
  });
});

describe('SHOP_TIMEZONES', () => {
  it('contains the default at +480', () => {
    expect(shopTimezoneById(DEFAULT_SHOP_TIMEZONE_ID)?.offsetMinutes).toBe(480);
  });

  it('has unique ids and in-range offsets', () => {
    const ids = new Set(SHOP_TIMEZONES.map((t) => t.id));
    expect(ids.size).toBe(SHOP_TIMEZONES.length);
    for (const tz of SHOP_TIMEZONES) {
      expect(tz.offsetMinutes).toBeGreaterThanOrEqual(-720);
      expect(tz.offsetMinutes).toBeLessThanOrEqual(840);
    }
  });

  it('mirrors the Dart catalog', () => {
    // Keep in lock-step with lib/core/utils/shop_timezones.dart.
    expect(SHOP_TIMEZONES.length).toBe(15);
  });
});

describe('formatShopDateTime', () => {
  it('composes date · time in the shop zone', () => {
    // Earlier tests in this file move the ambient zone — pin it back first.
    setAmbientShopTimezone({
      timezoneId: DEFAULT_SHOP_TIMEZONE_ID,
      offsetMinutes: DEFAULT_SHOP_OFFSET_MINUTES,
    });
    // 08:18Z is 4:18 PM in the default shop zone (Asia/Manila, +8).
    expect(formatShopDateTime(new Date('2026-09-01T08:18:00Z'))).toBe('Sep 1, 2026 · 4:18 PM');
  });
});
